File: src/main/java/com/donrobo/fpbg/blueprint/building/AbstractBuilding.java
package com.donrobo.fpbg.blueprint.building;

import net.sf.json.JSONObject;
import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;

public abstract class AbstractBuilding implements Building {
    private final int x;
    private final int y;

    public AbstractBuilding(int x, int y) {
        this.x = x;
        this.y = y;
    }

    @Override
    public int getX() {
        return x;
    }

    @Override
    public int getY() {
        return y;
    }

    @Override
    public JSONObject toJson() {
        JSONObject json = new JSONObject();

        json.put("name", getName());
        JSONObject posJson = new JSONObject();
        posJson.put("x", getX() + getBlueprintXOffset());
        posJson.put("y", getY() + getBlueprintYOffset());
        json.put("position", posJson);

        addCustomPropertiesToJson(json);

        return json;
    }

    protected void addCustomPropertiesToJson(JSONObject json) {
    }

    protected abstract double getBlueprintXOffset();

    protected abstract double getBlueprintYOffset();

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }

        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        AbstractBuilding that = (AbstractBuilding) o;

        return new EqualsBuilder()
                .append(getX(), that.getX())
                .append(getY(), that.getY())
                .isEquals();
    }

    @Override
    public int hashCode() {
        return new HashCodeBuilder(17, 37)
                .append(getX())
                .append(getY())
                .toHashCode();
    }
}

File: src/main/kotlin/com/donrobo/fpbg/blueprint/building/AbstractBuilding.kt
package com.donrobo.fpbg.blueprint.building

import com.donrobo.fpbg.blueprint.Direction
import com.donrobo.fpbg.data.Int2
import net.sf.json.JSONObject

abstract class AbstractBuilding(override val x: Int, override val y: Int) : Building {

    override fun toJson(): JSONObject {
        val json = JSONObject()

        json.put("name", name)
        val posJson = JSONObject()
        posJson.put("x", x + blueprintXOffset)
        posJson.put("y", y + blueprintYOffset)
        json.put("position", posJson)

        addCustomPropertiesToJson(json)

        return json
    }

    protected open fun addCustomPropertiesToJson(json: JSONObject) {}

    protected abstract val blueprintXOffset: Double

    protected abstract val blueprintYOffset: Double

    override val position: Int2
        get() = Int2(x, y)

    override fun move(offset: Int, direction: Direction): Building {
        val (x1, y1) = direction.move(Int2(0, 0), offset)

        return move(x1, y1)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other?.javaClass != javaClass) return false

        other as AbstractBuilding

        if (x != other.x) return false
        if (y != other.y) return false

        return true
    }

    override fun hashCode(): Int {
        var result = x
        result = 31 * result + y
        return result
    }
}
